FHEM/14_SD_UT.pm

```diff
@@ -1716,167 +1727,167 @@ sub SD_UT_Attr(@) {
 	my $bitData;
 	my $hex_lengh = length(InternalVal($name, "lastMSG", "0"));
 
-	if ($cmd eq "set" && $attrName eq 'repeats' && $attrValue !~ m/^[1-9]$/i) {
-		return "$name: Unallowed value $attrValue for the attribute repetition (must be 1 - 9)!";
-	}
-
-	############ chance device models ############
-	if ($cmd eq "set" && $attrName eq "model" && $attrValue ne $oldmodel) {
-
-		if (InternalVal($name, "bitMSG", "no data") ne "no data") {
-			my $devicemodel;
-
-			### ERROR for Users
-			my $allowed_models;
-			foreach my $keys (keys %models) {	# read allowed_models with the same hex_lengh
-				$allowed_models.= $keys.", " if ($models{$keys}{hex_lengh} eq $hex_lengh);
-			}
+	if ($cmd eq 'set') {
+		if ($attrName eq 'repeats' && $attrValue !~ m/^[1-9]$/xms) {
+			return "$name: Unallowed value $attrValue for the attribute repetition (must be 1 - 9)!";
+		}
+		if ($attrName eq 'UTfrequency' && ($attrValue !~ m/^[1-9]{1}[0-9]{0,2}\.?[0-9]*$/xms || $attrValue >= 1000.0)) {
+			return "$name: Invalid value $attrValue for the UTfrequency attribute. Values ​​such as: 433.92 are permitted.";
+		}
 
-			Log3 $name, 4, "SD_UT_Attr Check for the change, $oldmodel hex_lengh=$hex_lengh, attrValue=$attrValue needed hex_lengh=".$models{$attrValue}{hex_lengh};
-			return "ERROR! You want to choose the $oldmodel model to $attrValue.\nPlease check your selection.\nThe length of RAWMSG must be the same!\n\nAllowed models are: $allowed_models" if ($models{$attrValue}{hex_lengh} !~ /$hex_lengh/ && $oldmodel ne "unknown");	# variants one
-			return "ERROR! You want to choose the unknown model to $attrValue.\nPlease check your selection.\nRAWMSG length is wrong!\n\nAllowed models are: $allowed_models" if (not ($models{$attrValue}{hex_lengh} =~ /($hex_lengh)/ ) && $oldmodel eq "unknown");				  # variants two/three
-			### #### #### ###
+		############ change device models ############
+		if ($attrName eq 'model' && $attrValue ne $oldmodel) {
+			if (InternalVal($name, "bitMSG", "no data") ne "no data") {
+				my $devicemodel;
 
-			if ($attrName eq "model" && $attrValue eq "unknown") {
-				readingsSingleUpdate($hash, "state", " Please define your model with attributes! ", 0);
-			}
+				### ERROR for Users
+				my $allowed_models;
+				foreach my $keys (keys %models) {	# read allowed_models with the same hex_lengh
```

Comment:
Wäre hier nicht ein grep besser geeignet?

```
@allowed_models_list = { $_{hex_length} == $hex_length) keys %models;
$allowed_models= join('.',@allowed_models_list);
```

Bei HexLength bin ich mal einfach davon ausgegangen, dass es sich hier um einen Zahlenwert und nicht um einen String handelt. Vielleicht liege ich da aber auch falsch :)

Reply:
Das gelingt mir jetzt nicht.
Egal, wie ich die Zeile:
`@allowed_models_list = { $_{hex_length} == $hex_length) keys %models;
`
umstelle, erhalte ich: "syntax error".
So wird sie dann akzeptiert:
`my @allowed_models_list = { $_{hex_lengh} == $models{hex_lengh} };`
aber bekomme dann die Ausgabe:
`Allowed models are: HASH(0x4695ea8)
`

Abgesehen davon, ist "hex_lengh" (das "t" fehlt überall im Code :-) ein String.

Reply:
OK das grep fehlt. Hab ich vergessen:(

Reply:
Hätte ich auch drauf kommen können :-)
Gelungen ist es mir trotz etlicher Versuche nicht :-( Nur ein Auszug:

```
				# @allowed_models_list = { $_{hex_length} == $hex_length) keys %models;
				# my @allowed_models_list = grep { $models{hex_lengh}{$_} eq $hex_lengh } keys %models;
				# my @allowed_models_list = grep { $models{$_} eq $hex_lengh } keys %models;
				# my @allowed_models_list = grep { $_{hex_lengh} == $hex_lengh } keys %models;
				# my @allowed_models_list = grep { $hex_lengh == $_->{'hex_lengh'} } %models; # unknown
				# %allowed_models_list = map { $hex_lengh == $_->{hex_lengh} } @%models;
				# @allowed_models_list = grep { $hex_lengh == $_{hex_lengh} } %models;
				# my @allowed_models_list = grep { _$hex_lengh , $models{$keys}{hex_lengh} };
```

Der Hash ist so aufgebaut:

```
my %models = (
	# keys(model) => values
	"Buttons_five" =>	{ "011111"	=> "1_fan_low_speed",
											"111111" 	=> "2_fan_medium_speed",
											"111101" 	=> "3_fan_high_speed",
											"101111" 	=> "light_on_off",
											"111110"	=> "fan_off",
											hex_lengh	=> "3",
											Protocol 	=> "P29",
											Typ				=> "remote"
										},
	"CAME_TOP_432EV" =>	{	"1110"		=> "left_button",
```

Reply:
Nah dran, aber knapp vorbei :(
Also wir holen alle `keys `von `%models`. Und jeder `Key `befindet sich dann in `$_`

Um also auf ein Element im Hash jetzt für den Vergleich zugreifen zu können verwenden wir
`$models{$_}
 
Ich würde mal auf folgendes tippen:
`@allowed_models_list = grep { $models{$_}{hex_length} == $hex_length) keys %models;

Reply:
Danke, mit einer kleinen Anpassung funktioniert das jetzt.
https://github.com/RFD-FHEM/RFFHEM/commit/6bcd23729b7860f280569aab7cac36d10b5955c2